### ISSUE TITLE
`{ScriptBuf,Builder}::push_int` should accept `i32`, not `i64`

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#1530]
Only 32-bit integers are allowed in Bitcoin script so `i64` is misleading and potentially dangerous.

Also to avoid having to convert lock times with suspiciously-looking `as` I suggest we have `push_csv(lock_time: relative::LockTime)` and `push_cltv(lock_time: absolute::LockTime)`.

**Comment by apoelstra** [comment]
concept ACK to both suggestions.

Note that if `push_int` takes an `i32` then we **need** separate methods to push locktimes, since locktimes are 32-bit *unsigned* integers (which get encoded with 5 bytes if they are large enough), not 32-bit *signed* integers (which always fit into 4 bytes). I think the original code takes `i64` only because that's what Core does, but I suspect that this discrepancy was the reason for Core to make the choice it did.

**Comment by Kixunil** [comment]
Oh, true. In that case I guess we should keep `i64` in internal API and expose separate methods that just convert?

Also I was unspecific, I meant for `push_csv` to push int and `OP_CSV OP_DROP` instructions.

**Comment by sanket1729** [comment]
Some miscellaneous points about `CScriptNum`.

- Normal `CScriptNum` operations are within the range `[-2^31 +1...2^31 -1]`, it also does not include `i32::MIN` value :( 
- As with bitcoin, there are a few notable exceptions. Namely, CSV and CLTV accept `CScriptNum` that can hold 5 byte integers, `[-2^39 +1...2^39 -1]`. 
- It is also helpful to have `push_int` support 5-byte to compare the result of adding two 4-byte values that can overflow. Pasting relevant comment from bitcoin core.

```
 * Numeric opcodes (OP_1ADD, etc) are restricted to operating on 4-byte integers.
 * The semantics are subtle, though: operands must be in the range [-2^31 +1...2^31 -1],
 * but results may overflow (and are valid as long as they are not used in a subsequent
 * numeric operation). CScriptNum enforces those semantics by storing results as
 * an int64 and allowing out-of-range values to be returned as a vector of bytes but
 * throwing an exception if arithmetic is done or the result is interpreted as an integer.
```

But, all the above said, think having an `i32` is the better route. If someone really needs long-duration-based timelocks, we can add support for it later. 

If we were super careful, it might be useful to define a `ScriptNum` type and enforce the exact semantics by removing the i32::MIN value.

**Comment by Kixunil** [comment]
Oh, crap. In that case I prefer having `ScriptNum`.

Regarding not following with `OP_CSV` `OP_DROP`, would it be useful to have two methods one that does push them and one not? Or if we're going to have just one, without opcodes, we should think of a name that doesn't suggest those will be pushed.

**Comment by apoelstra** [comment]
If we do this, and it enforces minimality (which I think it might, in light of #1527) then we need to call it something other than `ScriptNum`, because Core's `CScriptNum` does not use the `OP_n` opcodes are used.

**Comment by Kixunil** [comment]
Does Core even encode it? I don't feel it has to be the same even if it's named similarly. The only other reasonable names that come to my mind are `ScriptNumber`, `ScriptInt`, and `ScriptInteger`.

**Comment by apoelstra** [comment]
Yes, the entire point of the `CScriptNum` class is that it represents a script-encoded number.

**Comment by Kixunil** [comment]
Well, in our case it'll be more like to-be-script-encoded number.

**Comment by Kixunil** [comment]
I don't think it's *that* bad. How about we have a last call period until say Wednesday to make PRs for these things? And if anything turns out complicated feel free to refuse to review.

**Comment by apoelstra** [comment]
Sounds good to me. I would like to get this in for 0.30, and I also don't think it's a ton of work ... but I agree that if, by Wednesday, there's still a nonobvious amount of work to be done, let's punt.

**Comment by apoelstra** [comment]
Moving this to 0.31 -- I like this, it's not *super* hard, but we've got some other priorities and it seems like there's not a ton of pressure for this. Happy to re-prioritize if others disagree.

**Comment by Kixunil** [comment]
I'd still like to have methods for pushing times.

**Comment by tcharding** [comment]
Please see patch 2 of #1629

**Comment by tcharding** [comment]
> Normal CScriptNum operations are within the range [-2^31 +1...2^31 -1], it also does not include i32::MIN value :(

I cannot find where this minimum being one bigger than i32::MIN is enforced. I can only see it mention in code comments. I've looked in `interpreter.[h|cpp]` and `script.[h|cpp]`. Do you know where to find that?

Specifically `CScriptNum::getint()` includes `i32::MIN`.

**Comment by apoelstra** [comment]
`i32::MIN` can only be represented in 2's complement. ScriptNums use signed magnitude. There is no room for the value.

It is not possible that the docs are stale, no. It would be a hardforking change to change this.

**Comment by tcharding** [comment]
Thanks man, I didn't realize the encoding had a name. Our (Bitcoin's) implementation is pretty wasteful though huh, at times using a whole byte to show the sign.

For anyone else reading the value 1 << 31 (0x800_000) can be represented by 32 bits in twos-compliment but requires five bytes in the bitcoin script num encoding so it cannot be read back in.

**Comment by tcharding** [comment]
> Only 32-bit integers are allowed in Bitcoin script

With my new found understanding; one can create a `CScriptNum` from any i64 and encode it into a script (a script that cannot be read back in) so we should provide some way for doing this still via a direct push, right? Perhaps:
```rust

    /// Adds instructions to push an absolute lock time onto the stack.
    pub fn push_lock_time(self, lock_time: absolute::LockTime) -> Builder {
        self.push_encoded_int(lock_time.to_consensus_u32().into())
    }

    /// Adds instructions to push a sequence number onto the stack.
    pub fn push_sequence(self, sequence: Sequence) -> Builder  {
        self.push_encoded_int(sequence.to_consensus_u32().into())
    }

    /// Adds instructions to push a signed 32 bit integer onto the stack.
    ///
    /// Integers are encoded as little-endian signed-magnitude numbers, but there are dedicated
    /// opcodes to push some small integers.
    pub fn push_int(self, data: i32) -> Builder {
        self._push_int(data.into())
    }

    /// Adds instructions to push a signed 64 bit integer onto the stack.
    ///
    /// Script nums that encode using more than 4 bytes are unable to be read back in. If
    /// you are creating a script that you hope to be readable consider using `push_int`.
    pub fn push_potentially_unreadable_int(self, data: i64) -> Builder {
        self._push_int(data)
    }

    fn _push_int(self, data: i64) -> Builder {
        // We can special-case -1, 1-16
        if data == -1 || (1..=16).contains(&data) {
            let opcode = opcodes::All::from(
                (data - 1 + opcodes::OP_TRUE.to_u8() as i64) as u8
            );
            self.push_opcode(opcode)
        }
        // We can also special-case zero
        else if data == 0 {
            self.push_opcode(opcodes::OP_FALSE)
        }
        // Otherwise encode it as data
        else { self.push_encoded_int(data) }
    }

    /// Adds instructions to push an integer onto the stack without optimization.
    ///
    /// This uses the explicit encoding regardless of the availability of dedicated opcodes.
    fn push_encoded_int(self, data: i64) -> Builder {
        let mut buf = [0u8; 8];
        let n = ScriptNum(data);
        let len = n.encode(&mut buf);
        self.push_slice(&buf[..len])
    }
```

**Comment by apoelstra** [comment]
> With my new found understanding; one can create a `CScriptNum` from any i64 and encode it into a script (a script that cannot be read back in) so we should provide some way for doing this still via a direct push, right?

I don't think we should bother supporting this. I get that Bitcoin Core's `CScriptNum` type allows this, but their CScriptNum isn't a public thing (Core is not even a library) and they're doing it specifically to handle internal overflows.

If we have a publicly exposed CScriptNum equivalent I think you should only be allowed to use it to construct values that can actually be read as valid CScriptNums.

Similarly I don't think we should expose a way to produce a negative 0.

If people want to do this stuff in direct pushes, they can do it by hand.

**Comment by tcharding** [comment]
> If we have a publicly exposed CScriptNum equivalent I think you should only be allowed to use it to construct values that can actually be read as valid CScriptNums.

This week I'm learning a bunch about invariants. Does this mean that we are trying to uphold the invariant that the inner value is always readable if encoded? Should we not implement `core::ops::Add` for example but implement a custom `add` method that checks that the invariant holds still after the addition?

**Comment by apoelstra** [comment]
Lol. Well, given that Bitcoin's own `OP_ADD` does not even preserve this invariant...

More seriously, I don't think we should implement any arithmetic opcodes on `CScriptNum`. This structure is purely for encoding and users shouldn't be using it until the last second.

**Comment by tcharding** [comment]
Oh that is interesting, so to paraphrase to see I get it, its _not_ a general purpose integer wrapper with weird encoding rules but a type that only  supports the encoding/decoding of specific script nums.

**Comment by apoelstra** [comment]
Yeah, that's the way I think of it.

Honestly, given that we have `Script::push_int` and `Script::push_int_non_minimal`, I'm not sure this type even needs to be publicly exposed.

**Comment by Kixunil** [comment]
We need to handle attempt to push `i32::MIN` somehow. Silently pushing it is dangerous at least. So either the method should return `Result` or we should have `ScriptNum` enforcing valid range. The advantage of the latter is we can still have `From<u8>` for instance so that at least some cases may be statically proven correct without unwrapping. Also validation can be naturally pushed to application boundary making error handling much simpler.

**Comment by apoelstra** [comment]
My vote now is to:

* Make `ScriptInt` non-public and only used internally to assist with the `push_int` logic
* Replace `push_int` with a bunch of `push_num_{i,u}{8,16,32,64,size}` and return `Result` from all but the 8s and 16s.

I'd also be fine with a single `push_num` which took a `I: Into<i64>` and returned a `Result` even though some types of inputs are statically guarantneed not to fail.

I suspect that the vast majority of uses for pushing numbers will be to push `usize` values which are, in practice, less than 20. So I wouldn't mind also an `_unchecked` variant of these which just encoded the numbers whether the numeric opcodes will accept them or not.

**Comment by Kixunil** [comment]
> I suspect that the vast majority of uses for pushing numbers will be to push `usize` values which are, in practice, less than 20.

Can you explain this a bit? To my knowledge integers are not very useful in Bitcoin script and not very used. Except for locks but I consider those special.

**Comment by sanket1729** [comment]
I think Andrew means pushing for `k` and `n` in Multi-sig scripts. Most of which are less than 20.

> Replace push_int with a bunch of push_num_{i,u}{8,16,32,64,size} and return Result from all but the 8s and 16s.

I am in favor of this, most users will use the API that does not fail. Only advance users need to worry about `i32::MIN` case which would be handled in `push_i32` as Result

**Comment by apoelstra** [comment]
> I think Andrew means pushing for `k` and `n` in Multi-sig scripts. Most of which are less than 20.

Yes, and perhaps weighted thresholds or something like that.

> To my knowledge integers are not very useful in Bitcoin script and not very used

Your knowledge is correct :)

**Comment by apoelstra** [comment]
This proposal sounds good to me.
